CMakePresets.json

```diff
@@ -254,11 +254,16 @@
       "inherits": ["linux-base"],
       "hidden": true,
       "environment": {
+        "CFLAGS": "$penv{CFLAGS}",
         "CXXFLAGS": "$penv{CXXFLAGS} -msimd128 -msse4.2",
         "LDFLAGS": "$penv{LDFLAGS} -msimd128 -msse4.2 -Wno-limited-postlink-optimizations -sALLOW_MEMORY_GROWTH=1 -sALLOW_TABLE_GROWTH=1"
       },
       "cacheVariables": {
-        "CMAKE_TOOLCHAIN_FILE": "$penv{EMSDK}/upstream/emscripten/cmake/Modules/Platform/Emscripten.cmake"
+        "CMAKE_TOOLCHAIN_FILE": "$penv{EMSDK}/upstream/emscripten/cmake/Modules/Platform/Emscripten.cmake",
+        "EMSCRIPTEN_SYSTEM_PROCESSOR": "",
```

Comment:
Note that I tried a bajillion different things to try to prevent needing this and spent far too long trying to figure out the proper way to check if a variable is uninitialized in Cmake. As far as I can tell, the behavior of the below idiom changed between v3.28.2 (on my Fedora 40 box) and v3.29 on the build machines. 

`if ("${EMSCRIPTEN_SYSTEM_PROCESSOR}" STREQUAL "")`

It blows up with an error (even when I disable --warn-uninitialized) and on my machine it works perfectly regardless of whether --warn-uninitialized is set or not.

I was unable to find any reference in the docs as to if this actually changed in 3.29, and I tried vendoring the toolchain and was completely unable to find any documentation as to how to detect an uninitialized variable in CMake. The most obvious: 

`if (NOT DEFINED EMSCRIPTEN_SYSTEM_PROCESSOR OR "${EMSCRIPTEN_SYSTEM_PROCESSOR}" STREQUAL "")`

Still explodes. So. This will work, I think. (at least it works in Act!)

Good enough for CI???? :worried:

Reply:
Interesting. I just updated to CMake 3.29 from 3.28 and can confirm that `--warn-uninitialized` catches these cases now. It should rightfully do so as well, since the proper way to check for uninitialized variables is using `NOT DEFINED SOME_VARIABLE`, since `"${SOME_VARIABLE}" STREQUAL ""` would catch variables that are technically initialized as well, as you seem to have found out already.

> It blows up with an error in the CI CMake (even when I disable --warn-uninitialized)

I believe `--warn-uninitialized` is implied when passing `-Werror=dev`, so you would need to disable/remove that one as well. I get no warning when both of those are disabled.

The real problem here is of course that we can't tell CMake to ignore warnings in the toolchain file, and it obviously can't know whether we wrote it ourselves or not.

Honestly I'd be down to just disable both of these flags instead. Nobody ever enables them anyway, as evidenced by none of the Emscripten authors catching this. Even if we define these particular variables someone is going to use more uninitialized variables down the line and break the CI or whatever.

Reply:
I just merged a new `strict` parameter to the `cmake-configure` action in #905, so you should be able to just do `strict: false` for web builds and you won't need these empty variables anymore.

Reply:
Unfortunately this doesn't seem to work as expected. If I pass strict: false, it still errors.

I actually would prefer fixing this "correctly" though, and since we already have to hack around various problems with the toolchain file, I just vendored it.

`if (NOT DEFINED VAR)` was enough, without the extra or is empty. Turns out CMAKE does not short-circuit, so that's why my compound if wasn't working. :banana: 

Vendoring also lets me clean up the various hacky things I added all over the place, and feels overall much more clean.

I will see if I can get a PR through to fix this upstream, but in the meantime...

Reply:
> If I pass strict: false, it still errors.

This doesn't make much sense to me. Plenty of projects are littered with use of uninitialized variables. I can't imagine CMake decided to suddenly make that a hard error without something enabling `--warn-uninitialized` and/or `-Werror=dev` somewhere somehow.

I am seeing mentions that `--warn-uninitialized` does in fact only check files that are at or under the project root (`CMakeLists.txt`), so I think part of the problem here is that the `EMSCRIPTEN_CACHE_FOLDER` is *in* the checkout folder as opposed to next to it (or above it) in the folder tree. Maybe there's another folder that could be used to house it, like `${{ runner.temp }}`? Although I don't know how that plays with the persistent cache.

> Vendoring also lets me clean up the various hacky things I added all over the place, and feels overall much more clean.

I don't love the idea of vendoring this, to be honest. I would have preferred the previous hacks, assuming that the uninitialized variable weirdness can be resolved in some way.

Reply:
> Maybe there's another folder that could be used to house it

I can explore this possibility. I think I can set the cache folder to anything.

 > I don't love the idea of vendoring this, to be honest.

I don't either, though considering the staleness of the PR fixing shared library builds, I do wonder if the Cmake support is actually maintained within Emscripten.

I can switch back to the hackified version. Worst case, I got it to work by stubbing all the variables out to empty string, and that remains viable. (though icky...)

Your checkout folder idea sounds promising though, so I'll try that first.

Reply:
Alas, the checkout folder idea didn't pan out. The Emscripten action does not appear to work with anything outside the source root. 

I might be missing something, but I think the workaround Good Enough(TM) for now.

Reply:
I just tried running the CI in regular GitHub Actions (not Act) without the CMake variables defined in and everything configures and builds just fine, with your latest commit, no errors no nothing.

Reply:
Huh. That's weird... I started the whole journey because CI was going boom here on Github, and used Act to help me figure it out. Maybe you did something slightly differently?

Here is an example failure from the job history: https://github.com/godot-jolt/godot-jolt/actions/runs/9571585844/job/26389026062#step:4:114

Oh wait, do you mean AFTER the strict mode change? I don't think I ever tried that outside Act... Hmmmmm...

Reply:
> Oh wait, do you mean AFTER the strict mode change?

Well, yes, technically, but I left strict mode enabled. I can see the respective CMake flags being passed to it.

Here's the run, assuming you have permission to see it: https://github.com/mihe/godot-jolt/actions/runs/9653686082

Reply:
I think it is exactly what you suspected, and since it wasn't cached, the folder is outside the build root and so the check doesn't happen.

Your toolchain file:
` CMAKE_TOOLCHAIN_FILE="/home/runner/work/_temp/ee5c6111-c45f-4c92-9497-edd8179428ec/emsdk-main/upstream/emscripten/cmake/Modules/Platform/Emscripten.cmake"`

This PRs, with working cache (due to it running more than once, I guess?):
`CMAKE_TOOLCHAIN_FILE="/home/runner/work/godot-jolt/godot-jolt/emsdk-cache/emsdk-main/upstream/emscripten/cmake/Modules/Platform/Emscripten.cmake"`

I am also suspecting now that the cache directory error I was seeing when trying to use the tmp folder was either an Act thing or my not setting the `needs:`

I will try again. Thanks for all your help on this!!

Reply:
:( 
`const fullCachePath = path.join(process.env.GITHUB_WORKSPACE, emArgs.actionsCacheFolder);`

Reply:
Honestly, judging by how quickly these jobs completed even without the cache being used, I would say just ditch `actions-cache-folder` thing altogether. Setting up Emscripten is never going to be a bottleneck for CI. That would then presumably make it use this temporary directory instead and fix all this.

Reply:
> What's a few more minutes when you have an hour-long CI. :)

Haha! Good point! :laughing: 

Done!